decisions: decide process, improve manpages

markus2330 opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you,
but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For docu fixes, spell checking, and similar none of these points below
need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://master.libelektra.org/doc/contrib/documentation.md))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

kodebach commented on doc/decisions/manpages.md:
```diff
@@ -11,17 +12,23 @@ Storing generated files is annoying, as it requires:
+3. [help2man](https://www.gnu.org/software/help2man/)
+4. Doxygen:
+   - Constraint 1 probably broken
+5. Pandoc: has quite a few dependencies and would need rewrite of the current documentation in doc/help:
```
The required dependencies [listed for Debian](https://packages.debian.org/bullseye/pandoc) all seem like pretty standard packages. The dependencies of the RPM package also seem very similar.

kodebach commented on doc/decisions/manpages.md:
```diff
@@ -11,17 +12,23 @@ Storing generated files is annoying, as it requires:
+   - Constraint 1 probably broken
+5. Pandoc: has quite a few dependencies and would need rewrite of the current documentation in doc/help:
+   - Constraint 1 fulfilled via https://pandoc.org/MANUAL.html#definition-lists
+   - would need YAML metadata for every file
+     (it is possible to also pass this as command-line arguments via `--variable` but then we would move meta-information about manpages to the build system)
```
First of all, the CMake code already needs to know which section a man page belongs to. We also use CMake to extract the proper date from the git log. So the build system is already somewhat involved.

To keep the information encoded in CMake scripts minimal, we could use a simple shell script that converts from

```markdown
# kdb-get(1) -- Get the value of a key stored in the key database
```

to

```markdown
---
title: KDB-GET
section: 1
...

## NAME

**kdb-get** - Get the value of a key stored in the key database
```

and pipe that output into `pandoc`. But I'd actually prefer just having the YAML front matter in the markdown file. Makes things more explicit and clear for beginners. It also get's rendered nicely as as table on GitHub. Random example I found: https://github.com/jonbeebe/frontmatter/blob/accd644b15a5ac7adaab261655eb552065020821/tests/testfile.md

markus2330 replied:
I did small clarifications in aec620f785ca38f1bdf0c954db9aaf59ed8a6c8a and added further possibilities to 5ee82b09cdd64a4192cb2daf6e128cc046aaae46

But as said: This PR was only to collect flying around information. Please add further information directly to the decision file after this was merged.

kodebach reviewed (commented):
IMO in a decision PR there should just be a single draft decision and the required release notes. But nothing else should change, because that complicates the reviews...

---

Regarding the man pages, I'm in favour of Pandoc. It's a pretty common tool and we could also use it to generate HTML for the website. Right now the man page looks different on the website, because it is just rendered as normal markdown without the special handling of ronn (e.g. the NAME section is missing).

markus2330 commented:
@hannes99 do you have time for a quick look?

@kodebach in 7de209ebfcba885059b4969b2fecd39fca4ee300 I wrote one more sentence in constraints:

- The person merging the decision PR must be someone else as the person that created the decision.

@flo91 can you also take a look, as we will use the process later in FLOSS.

kodebach commented:
> The person merging the decision PR must be someone else as the person that created the decision.

Sounds good :+1:

markus2330 commented:
[kodebach](https://github.com/kodebach) wrote in https://github.com/ElektraInitiative/libelektra/pull/4554#issuecomment-1292458985

> But until now the discussion wasn't completed, so there was no point in updating the document.

This is a grave misunderstanding of the decision process.

1. the discussion does not need to be "complete", on contrary: updates should be immediate. Different opinions should be reflected.
2. the foundations for a discussion (clear problem etc.) must be there **before** starting discussions about solutions

> To use a positive example from the past as illustration: In https://github.com/ElektraInitiative/libelektra/pull/4246#discussion_r1002442705 I suggested a new sentence to clear up some confusion. You said, "yep, that's good" and I knew and I can just add that information to the file.

No, this was me not following the process. The idea is to always add to the decision, and if it is non-working idea, it gets moved somewhere to alternative decisions or notes.

> AFAICT we addressed everything you asked for. Not in the decision file. That's true.

So we didn't address it at all for anyone who wants to understand the decision text.

> Now @atmaxinger should update the document and add using backendData->keys as an option.

No, the problem must be clear for everyone before. We also need more understanding about the interactions with other decisions.

I hopefully clarified everything above in 0c980ad0b10443c4f28a1be2ef89a30f29946af4

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -80,6 +84,8 @@ The first step is to create a PR with:
 - **one** decision, where at least the "Problem" and "Considered Alternatives" are filled out.
 - a link from [README.md](README.md) from the "Drafts" section to this decision.
 
+> At least the problem must be clear to everyone involved before the decision can leave the "Drafts" status.
```
I agree with that. But as you wrote 2 lines above: Just because the problem isn't clear to everyone, doesn't mean we can't add new "Considered Alternatives" already.

markus2330 replied:
Actually it is even required to have "Considered Alternatives". Any further clarifications needed?

kodebach commented:
> the discussion does not need to be "complete", on contrary: updates should be immediate.

Maybe "complete" wasn't the right word. Doing continuous updates is not the solution, unless the decision document should actually be full protocol of all discussions surrounding the decision. In that case, I misunderstood, but I'd also have to say you didn't state that very clearly.

My current understanding is that the decision document should reflect the contents of discussions, but not be an exact protocol. All arguments should be recorded. But only in as much detail as needed. I don't think there is a need to add 3 different explanations of the same argument, only because the first attempt didn't fully convince the decision author. It should be enough to summarize all attempts into one coherent argument and record that.

> Different opinions should be reflected.

Of course, but in this particular case, it wasn't even clear we had different opinions. Actually, in the end it turned out we agreed that my suggestion was a potential new approach. IMO it is not necessary to record the idea immediately, especially when the full consequences of the idea are not know yet.

When the discussion has progressed far enough that it is clear what the idea is and how it works, then we can record it together with any already existing arguments for/against it.

If we do it like you want to it would be:

1. Comment "there is also option X" is posted
2. Decision author copies the comment into the document to record it. But the author doesn't fully understand the idea, so asks for clarification.
3. Commenter provides clarification.
4. Decision author adds clarification to document. In that process they probably need to rewrite existing text from before.
5. Repeat 2-4 until all parties understand what "option X" is

This process just wastes a lot of time for the decision author, when they could just discuss in comments and then at then and add the option to the document. Yes, updating the document would record the changes in git. But do you honestly think anybody will look at the git history of a decision document to see how the explanation of an option changed?

> No, this was me not following the process. The idea is to always add to the decision, and if it is non-working idea, it gets moved somewhere to alternative decisions or notes.

So what would've been following the process? Me putting the sentence into the file and the asking if it's better? How would that be any different? Except that I'd have to spent a few more minutes opening an editor, editing a file, committing and pushing the change and only then posting a comment on GitHub to ask for your opinion.

> So we didn't address it at all for anyone who wants to understand the decision text.

I have now said **multiple** times. All we wanted to know is whether or not the additional information we suggested via comments was _enough_. Of course we would then add that to the document. Our problem was just that we _did not know_ what you were missing. Adding things at random is a bad idea. It may not help at all (i.e., waste of time) or worse it might lead to even more confusion.

> No, the problem must be clear for everyone before.

See https://github.com/ElektraInitiative/libelektra/pull/4562#discussion_r1006843090

---

Also you made it clear in 6a9f1d380a4c4f3c09e288805de0e82a11123416 that we don't want a benevolent dictatorship. And yet, you act very much like a benevolent dictator. The decision process works exactly as you say it works, not as the community understands it. The decision cannot be worked on, until you understand the problem (even though you rarely work on solutions).

atmaxinger commented on doc/decisions/decision_process.md:
```diff
@@ -24,13 +26,16 @@ Substantial decisions must be made in a transparent and participative way.
   - Prefer to directly give suggestions how to change sentences.
-  - General questions should be asked in the root of "Conversation" and not at random sentences in the review.
+  - General questions should be asked in the root of "Conversation" and not at vaguely related sentences in the review.
+- Decision PRs do not significantly change anything but one decision to make discussions clearly structured.
```
Funny how this very PR here is the opposite of this sentence. 😄

markus2330 replied:
Maybe the sentence was misunderstood, I removed to wrong rationale: bdb29bab5140c8ad406acdf1465f82b5776ab354
But yes, adding the man_pages here was definitely not a good idea.

atmaxinger commented on doc/decisions/decision_process.md:
```diff
@@ -64,6 +72,8 @@ We base our decision process and template on:
 Following subsections describe all steps a decision might run through.
 Only two of them are mandatory.
 
+In all steps we avoid discussions in issues or PRs but directly update the decision text with the different opinions and discuss the text of the decision PR.
```
How can we update the decision text, if we do not have any discussion about it? This kinda looks like a "throwing stuff on the wall to see what sticks" approach from the author's side.

atmaxinger commented on doc/decisions/man_pages.md:
```diff
@@ -0,0 +1,76 @@
+5. Pandoc:
+   - has a few standard dependencies
+   - would need rewrite of the current documentation in doc/help:
+     - To fulfill Constraint 1 [definition lists](https://pandoc.org/MANUAL.html#definition-lists) would be needed
```
Do we really need to rewrite the documentation, or would a "simple" `awk`/`sed` script be enough to fulfill this?

kodebach replied:
The rewrite process can definitely be automated. It's a one-time process, so we don't even need to use standard POSIX tools.

atmaxinger replied:
> It's a one-time process

In theory I agree, however we need to make sure that the updated markdown files with definition-lists syntax will also render correctly in at least:
- GitHub
- The Libelektra Website

kodebach replied:
AFAICT Github doesn't support definition lists. But I don't think that's a big deal. Few people will try to read the manpages on Github and except a perfect rendering. We could also use a different file extension instead of `.md` (e.g. `.pandoc`) then GitHub would just show the source and not try to render it.

For the website, we can simply use `pandoc` to generate HTML and use that.

flo91 commented on doc/decisions/man_pages.md:
```diff
@@ -0,0 +1,76 @@
+
+## Considered Alternatives
+
+0. `ronn-ng`, which doesn't have packages on most distribution (violates constraint 2.) and thus created this problem
```
At least it can be installed with `gem` (see https://github.com/apjanke/ronn-ng/blob/master/INSTALLING.md) which should be available on most distributions.
So I'm not sure if the availability of the tool should really be considered that bad.

markus2330 replied:
Many distributions require that they can build packages exclusively with packages they have. Is this different in Gentoo?

kodebach replied:
I think this just needs clarification that using `ronn-ng` **and** not committing the generated files is not possible because of this constraint. The current state on `master` is not a problem, unless we want to remove the generated files.

flo91 replied:
Hm, looks like I'm propably missing something here. :confused:

Packages that are offered by the package-managers of distributions usually must be created by a package maintainer.
Most distributions use pre-compiled binaries, here Gentoo is an exception where almost everything gets compiled locally.

So if package maintainers have to compile the software and package the binaries, is it a problem if they also must build the manpages in the process?
For developers and maintainers I think it should be ok if they have to install a software like ronn-ng manually.
Is this the point we are actually talking about here?
I've also read the discussion in https://github.com/ElektraInitiative/libelektra/issues/4542

End-users then get the pre-compiled packages with binaries and manpages from their distribution.

So the only real drawback I can see here if is someone wants
to build the package locally as mentioned in https://www.libelektra.org/installation/installation#os-independent.
But as compilation steps and installation of other utilities like debhelper is also necessary during
this process, a additional `gem install` doesn't look like a big problem to me.

But as I said, I'm propably missing something here, so I'd welcome clarification, thanks!

markus2330 replied:
Thx @kodebach the two interwoven problems were not explained well, I changed in f86628ce541a2a451eca393a758d083a643e0a59

@flo91 I gave an example about deb/dpkg but it would surprise me if it would be completely different in Gentoo. When you type `emerge elektra`, it might emerge other packages but it would be at least very bad style if `emerge elektra` itself would do a `gem install ronn-ng`. Would it be allowed that a `emerge ronn-ng` simply does a `gem install ronn-ng`? If this would be allowed, I have to admit that in Gentoo the `ronn-ng` problem is not a very though one. For most distributions, however, where alternative package installation methods are not allowed, packaging a ruby application is non-trivial. Hence most distribution don't have `ronn-ng` packaged. Debian actually has a [`ronn` packaged but it is an older version](https://packages.debian.org/sid/ronn). I clarified this in b5baca12a0ca875d518cd1d684654e92049d5ac8.

flo91 replied:
First of all, I never was a package-maintainer and never added a package to a distribution, so I'm definitely
not an expert in this topic, but I want to explain my understanding.

`ronn-ng` is basically a **build-time** dependency, not a run-time dependency.
So the package maintainers (who install `ronn-ng` alongside with the
required compilers, linkers, libs, etc.) could create the man pages and package them alongside with the compiled
binaries. The end-users who install the package, e.g. via `apt-get install libelektra`, don't need `ronn-ng` on
their local system because the package-maintainers already built the man pages.


So basically it would move the step of using `ronn-ng` to create the man pages to the package maintainer.
AFAIK in Gentoo you can also only define dependencies that are available via portage, but I think in this
case it should be also possible to package the already created man pages.
However, I'm focusing more on binary-based distributions like Debian here.



Your wrote in the decision text: 

> "But we want to avoid that distributions build packages without man pages, hence we added the generated files."

OK, so you don't want that the package **maintainers** have to install `ronn-ng` on their systems.
If that's the case, another possibility could be to create the man pages and add them to our repository for every release, but not in the `master` branch, so the building of man pages would be a part of the release process.
Package maintainers can then checkout the releases and get the appropriate man pages.

kodebach replied:
I'm also no expert, but I think markus2330 wanted to say that on Debian it must be possible to build the package on Debian without only packages available on Debian. At least, if we want the package to be in the Debian stable repo. Of course we can still build our own `.deb` packages as we do now.

> building of man pages would be a part of the release process

Only having the man pages as part of the release artifacts (maybe both as a separate `manpages.zip` and in the main release artifact) would be better than want we have now. But I don't know if it fulfils the Debian criteria. Maybe they just want to checkout a git commit and build from that. In that case the releases are not a solution.

markus2330 replied:
Exactly, like @kodebach says it does not really matter which kind of dependency it is. For integration into the archive all build and run-time dependencies must be present as packages. For Debian, and probably most other distributions, there must be a reproducible way how to compile a package (e.g. dpkg-buildpackage). And as part of that reproducibility, only packages can be allowed as any dependency.

As this is basics for maintainers and not really relevant for developers, I will not add it to the decision.

> If that's the case, another possibility could be to create the man pages and add them to our repository for every release, but not in the master branch, so the building of man pages would be a part of the release process.
Package maintainers can then checkout the releases and get the appropriate man pages.

Yes, this would be a different solution, with other drawbacks we can discuss when someone creates a decision PR for that.

kodebach replied:
> > If that's the case, another possibility could be to create the man pages and add them to our repository for every release, but not in the master branch, so the building of man pages would be a part of the release process.
> > Package maintainers can then checkout the releases and get the appropriate man pages.
> 
> Yes, this would be a different solution, with other drawbacks we can discuss when someone creates a decision PR for that.

Wouldn't this just be a "Considered Alternative" for this decision? It would solve the problem described here AFAICT.

flo91 replied:
Ok, thank you both for the helpful clarifications! :smiley:
After thinking about it, I understand that reproducibility of the build process with using only tools provided by the distribution can help the maintainers ensuring stability and maintainability.


> Wouldn't this just be a "Considered Alternative" for this decision? It would solve the problem described here AFAICT.

For me, it also looks more like a "Considered Alternative" than a whole new decision.

markus2330 replied:
As the TODO said I actually only wanted to clarify the problem, not the potential solutions, which should be done later by @hannes99 in a real decision PR. The update of the manpage decision in this PR was a fault on my side, it clearly showed to me that decision PRs should only contain one decision. :wink: 

I did an update nevertheless: 99cb07833852b9955e4690566e96f7ed404b4dab

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -24,13 +26,16 @@ Substantial decisions must be made in a transparent and participative way.
-  In general they also want change if it brings Elektra towards its [goals](/doc/GOALS.md) (but doesn't violate Elektra's stability guarantees).
+  In general people want change if it brings Elektra towards its [goals](/doc/GOALS.md) (but doesn't violate Elektra's stability guarantees).
 - We will always be able to reach an consensus.
   We don't need a vote (besides the approved review) or a benevolent dictatorship.
```
I read it as: 3 people agreeing **and** nobody else explicitly disagreeing.

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -3,14 +3,16 @@
 ## Constraints
 
 - All relevant information about decisions must be within Elektra's repository.
 - All decisions must go through at least two review rounds, with a merge in between.
```
I understood it as there should be two PRs for every decision. The first one introduces the problem and lists all options as considered alternatives. This will be merged after a review round. Then in a second PR one option is selected as "the decision" and this needs another review round.

flo91 replied:
Ok, that makes sense.
Is the 2nd PR with the selected option then already the actual implementation (code) PR or another text- and discussion-focused step, e.g. defining the selected option in more detail? In the 2nd case, we'd have a 3rd review round on the actual implementation, although one can argue that this is not a part of the actual decision process.

However, I think we should explain this part of the process in more detail, e.g. which types of PRs are intended and when they should be created.

markus2330 replied:
I clarified where the steps are explained in: e3ed16ee38c7a7de287321127df12611cd738ea7

It also defines what "clear" means, and improves structure (rationale about "unbiased moved down to rationale).

> Is the 2nd PR

There are up to 4 decision PRs possible, the 2nd could, even though not recommended, be already the one that cleans up the decision to move all details to the implementation. So I cannot really answer the question as the process does not define what the 2nd PR is. Could you please rephrase the questions with the steps explained in the decision process?

flo91 replied:
OK, so basically every step should be finished with merging a PR?
If that's the case, it should be stated in the description of the decision process.
For me it was not clear when to merge a PR and continue with a new PR and when it's recommended to push updates to an existing PR.

markus2330 replied:
Good point, this was actually not written anywhere. c28ec36f757131a2dccd34cfde05f66ca36501b0

markus2330 commented on doc/decisions/EXPLANATIONS.md:
```diff
@@ -83,10 +83,15 @@ This can be:
 Preferable it is linked, but if it is not possible, it can also be in full-text here.
 If particular information is important and not present in any sections above, please quote it here.
 
+Any discarded ideas and opinions that are not complete enough to be a "Considered Alternatives" can be written here.
```
I rephrased as requested and added an example. b26bc2e342246372c192f7fea8f274d1997f35c3

The example is quite general, if there are better ideas, please add a suggestion.

atmaxinger commented on doc/decisions/EXPLANATIONS.md:
```diff
@@ -83,10 +83,16 @@ This can be:
 If particular information is important and not present in any sections above, please quote it here.
 
+Any discarded, incomplete and unexplored ideas/opinions, which are not complete enough to be "Considered Alternatives", can be written here.
+For example, if it is obvious that the idea does not even solve the problem.
```
It would also be good to state WHY the idea is not complete enough/does not even solve the problem.

kodebach replied:
I also think it's not necessary to mention every idea that was thrown around. If an idea seems promising, but e.g. after digging through the code turns out to not solve anything, then yes we should mention it. But if an idea is flawed from the beginning, e.g. because of a logical error made by the person suggesting it, then I don't think we need to mention it. That might only confuse people reading the decision later on.

markus2330 replied:
> It would also be good to state WHY the idea is not complete enough/does not even solve the problem.

The main decision needs detailed rationale, the considered alternatives at least a short rationale why it was discarded. To also write rationale for notes is imho too much, I clarified in: 679609185c68de4d87245ae0e408d5b72af11b7d

kodebach replied:
I think the issue here is "discarded ideas". For "incomplete" and "unexplored" it there is no need for rationales. But when an idea is discarded, we need to record why. Otherwise, there will only be questions about it in future. Or worse someone will have to spend the time to rediscover why the idea was discarded.

markus2330 replied:
I agree, I removed "discarded ideas" as it is too general: 6b6687c63474a1bfbb6128d69480e5fc3a982a27

I expect notes about any discarded ideas (incomplete/unexplored) actually to be quite rare, at least it was rare up to now. To say more about it is IMHO already over-specification. I don't think that our goal is to have a bullet-proof decision meta-specification that specifies every possible content that might end up in a decision.

markus2330 commented on doc/decisions/decision_process.md:
```diff
@@ -70,6 +72,8 @@ We base our decision process and template on:
 Only two of them are mandatory.
 
+In all steps we avoid discussions in issues or PRs but directly update the decision text with the different opinions and discuss the text of the decision PR.
+
```
I rephrased the sentence 77a90b01bc2aee827391520c6a2c5824f06ff1e5 and removed a useless restriction in the notes.

markus2330 replied:
tl;dr

But whatever the problem is, there is a simple solution:
I added my constraints in 1c8d75fd5f665d3ba4f56e5c0827716c44db6d0b

Please make suggestions of what your constraints are so that you can participate in decisions (what I definitely want). Then we adapt the decision process to the constraints we personally have.

markus2330 commented:
> So what would've been following the process? Me putting the sentence into the file and the asking if it's better?

Yes, exactly. tl;dr about the rest. We both know that we cannot reach to an end if I make full replies. I don't want to take the time for that anymore. This is also one of the reasons to make the decision process strongly centered about decision text: so that single persons cannot fully occupy my time and the decision process becomes more technical. This is a personal decision and discussion about it is not respectful.

For personal problems, feel free to contact me per email and we can also have a short BBB call.

markus2330 commented on doc/decisions/EXPLANATIONS.md:
```diff
@@ -83,10 +83,15 @@ This can be:
 This section has links to other decisions with description what the relation is.
```
Excellent point! :revolving_hearts: 

I hope clarified it in 16e7d613bee15f3a33622164bcd7e11b875e70c3

flo91 replied:
Yes, that clarification looks good, thx!

markus2330 commented on doc/decisions/decision_process.md:
```diff
@@ -24,13 +26,16 @@ Substantial decisions must be made in a transparent and participative way.
 - People writing or reviewing decisions want Elektra to improve, so they also want to accept (acceptable) decisions.
-  In general they also want change if it brings Elektra towards its [goals](/doc/GOALS.md) (but doesn't violate Elektra's stability guarantees).
+  In general people want change if it brings Elektra towards its [goals](/doc/GOALS.md) (but doesn't violate Elektra's stability guarantees).
 - We will always be able to reach an consensus.
```
Actually we only need consensus for APIs, soon fixed in 1.0 anyway :rock:, everything else could be reimplemented for different needs. :rocket: 

@kodebach and I often have opinions like day and night and we always were able to come to an agreement :wink: 

I did clarifiations in 2e10d4b225fccb02bdb81e700efec6449354e638

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -3,14 +3,16 @@
-- At least two people need to be in favor of the decision in both rounds.
-- [Documentation guidelines](/doc/contrib/documentation.md) apply.
+- At least two people need to approve the decision in each round.
+- [Documentation guidelines](https://www.libelektra.org/devgettingstarted/documentation) apply.
```
Small note, but I'd prefer if we use relative links, where possible. That way people reading on GitHub won't be directed to the website, if the prefer GitHub. It also means you can use the link locally in your editor and don't need to first find out what file this is actually referring to. 

```suggestion
- [Documentation guidelines](../contrib/documentation) apply.
```

markus2330 replied:
This should be also clarified in documentation guidelines itself, can you please do an update there? Btw. I dislike `..` going to the root, this shouldn't be encouraged.

markus2330 commented on doc/decisions/decision_process.md:
```diff
@@ -64,6 +72,8 @@ We base our decision process and template on:
 Following subsections describe all steps a decision might run through.
 Only two of them are mandatory.
 
+In all steps we avoid discussions in issues or PRs but directly update the decision text with the different opinions and discuss the text of the decision PR.
```
> At least in early and volatile stages of the decision process, this approach sounds like generating quite some overhead.

Yes, I clarified: cb5b9129a68f0302894ed500a6c6bb1dbf11ad59

> 1.1. Maybe brainstorming techniques/sessions can help with finding more options.

If we (or someone) totally gets stuck, we might do this in a meeting. It never happened on a technical topic, though. So I didn't mention it. Furthermore, a meeting protocol would be needed, so that nothing important gets lost for the decision. Please give a concrete suggestion in the decision text about what would help you.

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -24,24 +26,42 @@ Substantial decisions must be made in a transparent and participative way.
+  Rather it is important that decisions:
+  - contain everything relevant, and
+  - what is written is technically correct.
+- Participants are explicitly allowed to skip discussions and only read the decisions text.
+  If the problem is not yet clear, only partial reviews are encouraged.
+  @markus2330 will do so of time budget reasons and to make sure the decisions stay understandable.
```
```suggestion
- Reviewers are only required to read the files in the PR.
  They may ignore the discussions surrounding the decision.
  Reviews should focus on the "Problem" section first, and only when that is agreed upon focus in the other parts.
  It is encouraged that at least one reviewer provides a review _without_ participating in the discussion.
  This ensures that there aren't any unintentional shared assumptions between discussion participants.
```

1. "clear" is not precise enough for me. "Clear" to who? I think agreement is a much better condition. This also leaves the option for reviewers to say "I don't fully understand it, but I can see that this could be a problem".
2. I don't think we should mention individual people in the decision.

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -24,24 +26,42 @@ Substantial decisions must be made in a transparent and participative way.
   So we do not expect that a clear problem or solution is in the decision writer's mind beforehand.
   Instead the decision process is a supported learning process.
+- People focus on getting the best solutions and not to wish for the impossible.
+- After 1.0, due to stability guarantees, not so many decisions are required anymore.
```
I'm not so sure about that. Things could be added as entirely new API without touching the existing parts. Why would that require less of a decision process? Also I think the stability guarantees would just _increase_ the need for a full decision process, if we want to change something. It is much harder to find a correct solution, if we have to think about stability as well.

markus2330 replied:
I agree, I completely changed that: c1a3d5c9341d24b6391c49cf098a237485ede18c

kodebach commented:
> so that single persons cannot fully occupy my time and the decision process becomes more technical.

Actually this is part of my problem with this decision. It is very focused on _you_. AFAICT you at least _want to_ review everything. If _your time_ is the limiting factor than we clearly have some form of benevolent dictator model. I'm fine with that too. It makes a lot of sense from the university POV. However, we shouldn't pretend it's not the case then.

I do respect that you have limited time and if adding everything to the file really is that much better for you, I'll try to do it. But you also have to respect _our_ limited time and should at least try to be helpful. In particular the problem from #4574 hasn't been addressed here at all. Both atmaxinger and I didn't understand _your review_. We understood that you were not satisfied with the current state, but we did not know what the problem was. We tried to get more information out of you, but failed. Maybe we used the wrong approach.

IMO a reviewer should be willing and even required to explain their opinion when asked to. I know you can do that, because you did it in https://github.com/ElektraInitiative/libelektra/pull/4574#discussion_r1000880243. "Don't like this" is a useless personal opinion that may be ignored IMO. But the explanation you gave when asked "I don't like that we get too much state/too many illegal sequences" is a valid concern that has to be taken into account.

With the "problem is unclear" issue it is not so easy. I admit that. Participating in a long discussion is too much to ask, yes. But, when the other party doesn't understand what is "unclear", then you have to give something. Either answer some questions, or at least explain what you did understand so there is some basis upon which to rewrite the text.

---

On a final note: This whole decision process does make sense in principle, but can be very off-putting to newcomers. Especially, when people come in to work on a particular topic and want to be done ASAP. I'm not entirely sure what could be done about this. For me personally, it's no longer an issue. I've given up and accepted that I'll be done, when I'm done.

atmaxinger commented on doc/decisions/decision_process.md:
```diff
@@ -74,9 +99,11 @@ The first step is to create a PR with:
 - **one** decision, where at least the "Problem" and "Considered Alternatives" are filled out.
```
What if there are no considered alternatives yet?

markus2330 replied:
Yes, you are right, I changed in 190cec69746f77ed9824bce3e01397076bbc1f6a

markus2330 commented on doc/decisions/decision_process.md:
```diff
@@ -24,24 +26,40 @@ Substantial decisions must be made in a transparent and participative way.
   - Prefer to directly give suggestions how to change sentences.
-  - General questions should be asked in the root of "Conversation" and not at random sentences in the review.
+  - General questions should be asked in the root of "Conversation" and not at vaguely related sentences in the review.
+- Decision PRs do not significantly change anything but one decision.
```
> I would even forbid typo fixes outside the already changed files.

I think at least backlinks should be allowed, otherwise it gets unpractical, see 5b70006d13907b6101ff60cd4eab5ffa0f8163a8

markus2330 commented on doc/decisions/decision_process.md:
```diff
@@ -24,24 +26,40 @@ Substantial decisions must be made in a transparent and participative way.
+- Decision PRs do not significantly change anything but one decision.
+- Changes not changing the decision step or the direction of the decision are not decision PRs.
+- The person merging the decision PR must be someone else as the person that created the decision.
+- The purpose of decisions is to have clear descriptions of technical problems and solutions.
```
decisions vs. decision process. I put it together in: 23ba841c7c544ffd828a340465f343f9062570c2

markus2330 commented:
> Actually this is part of my problem with this decision. It is very focused on you.

Yes, I started the documentation of the decision process because of concrete problems that I had (ideas floating around in many threads which I was unable to fully discuss and summarize in decisions anymore).

I clarified in 8762b6b93a721978a71477b4b26042d52e29deea

> AFAICT you at least want to review everything.

I think it is my duty to review everything from the students who I supervise. I hope that eventually there will be independent people (like you were&hopefully will be) that also find ways to significantly work on Elektra (e.g. by doing a PhD and supervising students or by founding a company related with CM). When this happens, I'll only review their stuff if I find it interesting, e.g., touches a part of Elektra my students or myself currently use or work on. The decision process was designed to support such scenarios.

> If your time is the limiting factor than we clearly have some form of benevolent dictator model. I'm fine with that too. It makes a lot of sense from the university POV. However, we shouldn't pretend it's not the case then.

This new process is specifically done as step to get away from the benevolent dictator model. We already had following models:

1. When I worked (nearly) alone on Elektra, I obviously decided everything. :wink: 
2. Also longer time ago, it used to be that students directly write to me for any problems, I gave them an answer and later we documented the decision (if at all).
3. Later, we discussed openly in issues (more transparency) but still the decisions were done usually only by me.

Only since we do decisions there were decisions which sometimes did not went to my favor. E.g. semantics_name.md was clearly against my opinion. Btw. this decision is written terribly, it needs some rework. Anyway, the ®elektra can stay, even though I don't really see a big use case for it. (And in the few use cases where it could have been used it was not used.) But please no discussion here other than about the process how to update such decisions. (I think it is clear, simply start a decision PR.)

Anyway, opinions are irrelevant, we want to do science here. I wrote the assumption in cb5d3116003755586f093282c46486d38d23a1a4

Now, I don't even fully read all discussions anymore and I think this is a good thing: I don't want to interfere everywhere all the time (which would also be some kind of dictatorship).

So I declare the end of the dictatorship :fireworks: 

> I do respect that you have limited time and if adding everything to the file really is that much better for you, I'll try to do it. But you also have to respect our limited time and should at least try to be helpful.

Yes, of course. The idea of this process that **everyone** saves time, see also last sentence.

> In particular the problem from https://github.com/ElektraInitiative/libelektra/pull/4574 hasn't been addressed here at all. Both atmaxinger and I didn't understand your review. We understood that you were not satisfied with the current state, but we did not know what the problem was. We tried to get more information out of you, but failed. Maybe we used the wrong approach.

> IMO a reviewer should be willing and even required to explain their opinion when asked to.

Yes, good point. Added it to assumptions: 32a4748ff657cf6fcf2be2c2e7d520116b2f0f3b

> With the "problem is unclear" issue it is not so easy. I admit that. Participating in a long discussion is too much to ask, yes. But, when the other party doesn't understand what is "unclear", then you have to give something. Either answer some questions, or at least explain what you did understand so there is some basis upon which to rewrite the text.

Yes, you can also ask about what I did understand.

> On a final note: This whole decision process does make sense in principle, but can be very off-putting to newcomers. 

I agree, I added the implication c579b76971d237754c0bca44937792436382254e

> Especially, when people come in to work on a particular topic and want to be done ASAP.

I hope that decisions increase the probability to get work done ASAP. It avoids dead-end implementations, weeks of tinkering around without feedback from experienced Elektra developers etc.

markus2330 commented:
We can make a 3rd round if requested at the meeting tomorrow.

kodebach commented on doc/decisions/decision_process.md:
```diff
@@ -129,23 +180,31 @@ These decision PRs are also merged for documentation purposes.
 - The decision process creates at least:
   - two chances to comment decisions, and
   - two commits in the git history.
+- It might be a barrier for newcomers to write a decision.
+  This is considered to be okay, as topics that need a decision are not the topics for newcomers.
```
I guess you added that because of my recent remarks... But it doesn't address my concerns.

Here "newcomers" should refer to "people writing their first decision" not necessarily "people contributing to Elektra the first time". However, there is a lot of overlap. Many of the recent big changes that require decisions (change-tracking, new mounting, new kdb tool, new spec) are being done by people new to Elektra.

I wasn't actually concerned with the extra effort of writing a decision. Good documentation should be encouraged. That takes time, no way around that. My issue was that decisions are such a dry and theoretical process. _That_ can be very off-putting, if you had an idea and wanted to write code.

kodebach replied:
Also the long and tedious process makes _me_ want to suggest fewer ideas, because I don't want to spend time on decisions.

markus2330 replied:
It is on purpose that less ideas get thrown around but instead we focus on decisions, which have much higher probability of actually getting done.

kodebach reviewed (commented):
A few final remarks...

kodebach commented:
Also just a note, but @markus2330 you technically broke the "some one other than the decision author must merge the PR" rule here.

kodebach commented:
One more note: The `doc/decisions/README.md` is really bad for merge conflicts. We almost can't have two separate decision PRs without merge conflicts. IMO we should 
- either use something like YAML frontmatter in the markdown files and thereby tag the files themselves.
- or use a folder structure e.g. `doc/decisions/drafts/...`, etc.

markus2330 commented:
@kodebach I like the idea with the folder structure. Can you explain your idea how to get the website https://www.libelektra.org/decisions/ nicely rendered (more or less like now)?

Alternatively, we could simply do what we do for release notes (use random line to add).

kodebach commented:
> Alternatively, we could simply do what we do for release notes (use random line to add).

That's also a hack and I've also suggested a long time ago to do release notes differently. But for now the random lines hack is good enough there.

> Can you explain your idea how to get the website https://www.libelektra.org/decisions/ nicely rendered (more or less like now)?

I'll make a PR.

kodebach commented:
> > Can you explain your idea how to get the website https://www.libelektra.org/decisions/ nicely rendered (more or less like now)?
> 
> I'll make a PR.

See #4632
